dice_ml/explainer_interfaces/dice_KD.py

```diff
@@ -0,0 +1,236 @@
+
+        # initializing model variables
+        self.model = model_interface
+
+        # loading trained model
+        self.model.load_model()
+
+        # variables required to generate CFs - see generate_counterfactuals() for more info
+        self.cfs = []
+        self.features_to_vary = []
+        self.cf_init_weights = []  # total_CFs, algorithm, features_to_vary
+        self.loss_weights = []  # yloss_type, diversity_loss_type, feature_weights
+        self.feature_weights_input = ''
+        self.hyperparameters = [1, 1, 1]  # proximity_weight, diversity_weight, categorical_penalty
+        self.optimizer_weights = []  # optimizer, learning_rate
+
+        # number of output nodes of ML model
+        temp_input = tf.convert_to_tensor([tf.random.uniform([len(self.data_interface.encoded_feature_names)])], dtype=tf.float32)
+        self.num_ouput_nodes = self.model.get_output(temp_input).shape[1]
+
+        # Stores the predictions on the training data
+        #TODO: Implement this for private dataset instance too
+        dataset_dict = self.data_interface.prepare_dataset_instance(encode=True)
+        dataset_dict_output = np.array([dataset_dict.values], dtype=np.float32)
+
+        self.predictions = self.predict_fn(dataset_dict_output)
+        self.predictions_vals = np.reshape(self.predictions[0], (self.predictions.shape[1],))
+
+        #segmenting the dataset according to outcome
+        #TODO: automate this for each outcome
+        dataset_with_predictions_all = self.data_interface.append_predictions(self.predictions_vals)
+        self.dataset_with_predictions = {}
+        self.dataset_with_predictions[1] = dataset_with_predictions_all.loc[np.round(dataset_with_predictions_all[self.data_interface.outcome_name + "_pred"])==1]
+        self.dataset_with_predictions[0] = dataset_with_predictions_all.loc[np.round(dataset_with_predictions_all[self.data_interface.outcome_name + "_pred"])==0]
+
+        # Prepares the KD trees for DiCE - 1 for each outcome (here only 0 and 1)
+        #TODO: make separate KD trees for each outcome
+        self.KD_tree = {
+            1: KDTree(pd.get_dummies(self.dataset_with_predictions[1][self.data_interface.feature_names])),
+            0: KDTree(pd.get_dummies(self.dataset_with_predictions[0][self.data_interface.feature_names]))}
+
+    def generate_counterfactuals(self, query_instance, total_CFs, desired_class="opposite", proximity_weight=0.5, diversity_weight=1.0, categorical_penalty=0.1, algorithm="DiverseCF", features_to_vary="all", permitted_range=None, yloss_type="hinge_loss", diversity_loss_type="dpp_style:inverse_dist", feature_weights="inverse_mad", optimizer="tensorflow:adam", learning_rate=0.05, min_iter=500, max_iter=5000, project_iter=0, loss_diff_thres=1e-5, loss_converge_maxiter=1, verbose=False, init_near_query_instance=True, tie_random=False, stopping_threshold=0.5, posthoc_sparsity_param=0.1, posthoc_sparsity_algorithm="linear"):
```

Comment:
among the optional parameters, only include the parameters that you need. Remove all else.